hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -129,52 +166,83 @@ def _get_provider(backend: str) -> Union[type, PrimitiveProvider]:
+        "shrink-phase": NotRequired[PhaseStatistics],
+        "stopped-because": NotRequired[str],
+        "targets": NotRequired[Dict[Optional[str], float]],
+    },
+)
+
+
 class ConjectureRunner:
     def __init__(
         self,
-        test_function,
+        test_function: Callable[[ConjectureData], None],
         *,
-        settings=None,
-        random=None,
-        database_key=None,
-        ignore_limits=False,
-    ):
-        self._test_function = test_function
-        self.settings = settings or Settings()
-        self.shrinks = 0
-        self.finish_shrinking_deadline = None
-        self.call_count = 0
-        self.valid_examples = 0
-        self.random = random or Random(getrandbits(128))
-        self.database_key = database_key
-        self.ignore_limits = ignore_limits
+        settings: Settings | None = None,
+        random: Random | None = None,
+        database_key: bytes | None = None,
+        ignore_limits: bool = False,
+    ) -> None:
+        self._test_function: Callable[[ConjectureData], None] = test_function
+        self.settings: Settings = settings or Settings()
+        self.shrinks: int = 0
+        self.finish_shrinking_deadline: float | None = None
+        self.call_count: int = 0
+        self.valid_examples: int = 0
+        self.random: Random = random or Random(getrandbits(128))
+        self.database_key: bytes | None = database_key
+        self.ignore_limits: bool = ignore_limits
 
         # Global dict of per-phase statistics, and a list of per-call stats
         # which transfer to the global dict at the end of each phase.
-        self._current_phase = "(not a phase)"
-        self.statistics = {}
-        self.stats_per_test_case = []
+        self._current_phase: str = "(not a phase)"
+        self.statistics: StatisticsDict = {}
+        self.stats_per_test_case: list[CallStats] = []
 
-        self.interesting_examples = {}
+        self.interesting_examples: dict[InterestingOrigin | None, ConjectureResult] = {}
         # We use call_count because there may be few possible valid_examples.
-        self.first_bug_found_at = None
-        self.last_bug_found_at = None
+        self.first_bug_found_at: int | None = None
+        self.last_bug_found_at: int | None = None
 
-        self.shrunk_examples = set()
+        self.shrunk_examples: set[InterestingOrigin | None] = set()
 
-        self.health_check_state = None
+        self.health_check_state: HealthCheckState | None = None
 
-        self.tree = DataTree()
+        self.tree: DataTree = DataTree()
 
-        self.provider = _get_provider(self.settings.backend)
+        self.provider: type | PrimitiveProvider = _get_provider(self.settings.backend)
 
-        self.best_observed_targets = defaultdict(lambda: NO_SCORE)
-        self.best_examples_of_observed_targets = {}
+        self.best_observed_targets: dict[str | None, float] = defaultdict(
+            lambda: NO_SCORE
+        )
```

Comment:
That's not a dict, it's a `MutableMapping` more specifically a defaultdict.  

(and I think this matters due to the invariance of mutable collections)